### ISSUE TITLE
The client never stops to request the server even if the file has been completely dowloded

### ISSUE THREAD

**Issue by GoogleCodeExporter** [bdshadow/gwtupload#47]
```
What steps will reproduce the problem?
1.launch the upload widget in simulation mode
2.load any file

the download stops but the status requests do not stop, because the 
finished message is not received.

I use GwtUpload5.8, windows vista

I've got a patch that consists in adding a save() call in the method 
setFinished of the class AbstractUploadListener. 
As a matter of fact, in the AppEngine with cache management, the finished 
boolean is not updated in the cache, so the status request from the client 
never see the finished boolean to true.

Raoul


```

Original issue reported on code.google.com by `root....@gmail.com` on 1 May 2010 at 9:59
- Merged into: #45

**Comment by GoogleCodeExporter** [comment]
```
Does it happen in GAE or also in a standard servlet container.?
What is 'simulation mode', you mean GAE in eclipse?
If you use GAE, have you added gwtupload-gae library to your project and are you
using AppEngineUploadAction?


-Manolo

```

Original comment by `manuel.carrasco.m` on 3 May 2010 at 6:35
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
I got the same problem, and analysing source code hunting for a bug, I found the
following: UploadServlet has a method 

protected void finish(HttpServletRequest request) {
    AbstractUploadListener listener = getCurrentListener(request);

    if (listener != null) {
      listener.setFinished(true);
    }
  }

where listener object is changing, but actually not... what I mean is.. just 
adding this:

  protected void finish(HttpServletRequest request) {
    AbstractUploadListener listener = getCurrentListener(request);

    System.out.println( "current listener before changing: " + listener );

    if (listener != null) {
      listener.setFinished(true);
      System.out.println( "listener changed: " + listener );
    }

    System.out.println( "current listener after changing: " +
getCurrentListener(request) );
  }

My console has the following written:

current listener before changing: counter=10 total=2229 done=2229 
cancelled=false
finished=false saved=Fri May 14 07:41:22 UTC 2010
listener changed: counter=10 total=2229 done=2229 cancelled=false finished=true
saved=Fri May 14 07:41:22 UTC 2010
current listener after changing: counter=10 total=2229 done=2229 cancelled=false
finished=false saved=Fri May 14 07:41:22 UTC 2010

as you see setting "listener.setFinished(true);" has no effect on current 
listener in
memory in case of GAE. Seems that's the root of problem. Keep investigating how 
to
fix that (I used both 5.8 and 6.0 versions - they both have this defect).
```

Original comment by `gdata.yu...@gmail.com` on 14 May 2010 at 8:13

**Comment by GoogleCodeExporter** [comment]
Original comment by `manuel.carrasco.m` on 17 May 2010 at 6:21
- Changed state: **Duplicate**
